File: src/main/java/org/rgbit/rgpg/Generate.java
package org.rgbit.rgpg;

public class Generate {
  private StringTrans trans = StringTrans.maketrans(Essential.TRANS_IN, Essential.TRANS_OUT);

  public String generate(String input, int length) {
    String md5K1 = MD5Helper.md5(Essential.K1 + input);
    String md5K2 = MD5Helper.md5(Essential.K2 + md5K1);
    String md5K3 = MD5Helper.md5(Essential.K3 + md5K1);

    StringBuilder sb = new StringBuilder();
    for (int i = 0; i < md5K1.length(); ++i) {
      if (md5K2.charAt(i) < md5K3.charAt(i)) {
        sb.append(md5K3.charAt(i));
      } else {
        sb.append(md5K2.charAt(i));
      }
    }
    String result = trans.translate(sb.toString());

    return Essential.FIRST + result.substring(1, length);
  }
}

File: rgpg/generate.py
# -*- coding: utf-8 -*-

"""
Implement password generator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:copyright: (c) 2018 by rgb-24bit.
:license: MIT, see LICENSE for more details.
"""

from hashlib import md5


def md5_hexdigest(a, b):
    """MD5 `a + b` and return hexdigest str."""
    return md5((a + b).encode('utf-8')).hexdigest()


def generate(key, essential, length=16):
    """Password generator."""
    md5k1 = md5_hexdigest(essential.K1, key)
    md5k2 = md5_hexdigest(essential.K2, md5k1)
    md5k3 = md5_hexdigest(essential.K3, md5k1)

    password = ''.join(map(max, zip(md5k2, md5k3)))

    table = str.maketrans(essential.TRANS_SRC, essential.TRANS_DST)
    return essential.FIRST + password[1: min(length, 32)].translate(table)
